docs/x509/reference.rst

```diff
@@ -388,6 +388,93 @@ X.509 CRL (Certificate Revocation List) Object
 
         The extensions encoded in the CRL.
 
+X.509 Certificate Builder
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. class:: CertificateBuilder
+
+    .. method:: issuer_name(name)
+
+        Sets the issuer's distinguished name.
+
+        :param name: The :class:`~cryptography.x509.Name` that describes the
+            issuer (CA).
+
+    .. method:: subject_name(name)
+
+        Sets the subject's distinguished name.
+
+        :param name: The :class:`~cryptography.x509.Name` that describes the
+            subject.
+
+    .. method:: public_key(public_key)
+
+        Sets the subject's public key.
+
+        :param public_key: The subject's public key. This can be one of
+            :class:`~cryptography.hazmat.primitives.asymmetric.rsa.RSAPublicKey`,
+            :class:`~cryptography.hazmat.primitives.asymmetric.dsa.DSAPublicKey` or
+            :class:`~cryptography.hazmat.primitives.asymmetric.ec.EllipticCurvePublicKey`
+
+    .. method:: serial_number(serial_number)
+
+        Sets the certificate's serial number (an integer).  The CA's policy
+        determines how it attributes serial numbers to certificates.  The only
+        requirement is that this number uniquely identify the certificate given
+        the issuer.
```

Comment:
We should probably mention the "at least 20 minutes of entropy" line, and
maybe link to our secure random number generation page? (Which should
probably document how to get an itneger... I'll file a ticket on that)

On Sun, Aug 2, 2015 at 4:13 PM, Paul Kehrer <notifications@github.com>
wrote:

> In docs/x509/reference.rst
> <https://github.com/pyca/cryptography/pull/2098#discussion_r36047152>:
>
> > +
> > +    .. method:: public_key(public_key)
> > +
> > +        Sets the subject's public key.
> > +
> > +        :param public_key: The subject's public key. This can be one of
> > +            :class:`~cryptography.hazmat.primitives.asymmetric.rsa.RSAPublicKey`,
> > +            :class:`~cryptography.hazmat.primitives.asymmetric.dsa.DSAPublicKey` or
> > +            :class:`~cryptography.hazmat.primitives.asymmetric.ec.EllipticCurvePublicKey`
> > +
> > +    .. method:: serial_number(serial_number)
> > +
> > +        Sets the certificate's serial number (an integer).  The CA's policy
> > +        determines how it attributes serial numbers to certificates.  The only
> > +        requirement is that this number uniquely identify the certificate given
> > +        the issuer.
>
> UUID4 isn't a bad choice converted to int. Current CABForum baseline
> requirements say "CAs SHOULD generate non‐sequential Certificate serial
> numbers that exhibit at least 20 bits of entropy". You can also do things
> like microtime + urandom to get to 160 bits
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pyca/cryptography/pull/2098/files#r36047152>.
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084